### PR TITLE
NIFI-2943 - pkcs12 keystore improvements

### PR DESCRIPTION
**Pull request by brosander**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:
### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
   in the commit message?
- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?
- [x] Is your initial contribution a single, squashed commit?
### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] N/A - If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] N/A - If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [x] N/A - If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] N/A - If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?
### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?
### Note:

Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
1. loading pkcs12 keystores with bouncy castle everywhere
2. tls-toolkit client using jks truststore when keystore type is specified differently
3. tests

### CODE REVIEW

**Comment by ishiyayu** [comment]
2016/10/28 午前6:32 "Bryan Rosander" <notifications@github.com>:

File: nifi-commons/nifi-security-utils/src/main/java/org/apache/nifi/security/util/KeyStoreUtils.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    public static String getKeyStoreProvider(String keyStoreType) {
+        if (KeystoreType.PKCS12.toString().equalsIgnoreCase(keyStoreType)) {
+            return BouncyCastleProvider.PROVIDER_NAME;
+        }
+        return null;
+    }
+
+    public static KeyStore getKeyStore(String keyStoreType) throws KeyStoreException {
+        String keyStoreProvider = getKeyStoreProvider(keyStoreType);
+        if (StringUtils.isNoneEmpty(keyStoreProvider)) {
```
Comment (alopresto) [inline nifi-commons/nifi-security-utils/src/main/java/org/apache/nifi/security/util/KeyStoreUtils.java]: I think this should be `StringUtils.isNotEmpty` -- the current method works but is designed to iterate over a number of `CharSequences`; hence the jarring name.

**Comment by alopresto** [comment]
I think we should provide log output indicating that the user's choice of PKCS12 is not used for truststores. 

```
hw12203:...assembly/target/nifi-toolkit-1.1.0-SNAPSHOT-bin/nifi-toolkit-1.1.0-SNAPSHOT (pr1165) alopresto
🔓 46s @ 19:28:29 $ ./bin/tls-toolkit.sh standalone -n 'localhost' -T PKCS12 -P password -S password
2016-11-14 19:52:11,629 INFO [main] o.a.n.t.t.s.TlsToolkitStandaloneCommandLine No nifiPropertiesFile specified, using embedded one.
2016-11-14 19:52:11,956 INFO [main] o.a.n.t.t.s.TlsToolkitStandalone Running standalone certificate generation with output directory ../nifi-toolkit-1.1.0-SNAPSHOT
2016-11-14 19:52:12,407 INFO [main] o.a.n.t.t.s.TlsToolkitStandalone Generated new CA certificate ../nifi-toolkit-1.1.0-SNAPSHOT/nifi-cert.pem and key ../nifi-toolkit-1.1.0-SNAPSHOT/nifi-key.key
2016-11-14 19:52:12,408 INFO [main] o.a.n.t.t.s.TlsToolkitStandalone Writing new ssl configuration to ../nifi-toolkit-1.1.0-SNAPSHOT/localhost
2016-11-14 19:52:13,382 INFO [main] o.a.n.t.t.s.TlsToolkitStandalone Successfully generated TLS configuration for localhost 1 in ../nifi-toolkit-1.1.0-SNAPSHOT/localhost
2016-11-14 19:52:13,382 INFO [main] o.a.n.t.t.s.TlsToolkitStandalone No clientCertDn specified, not generating any client certificates.
2016-11-14 19:52:13,382 INFO [main] o.a.n.t.t.s.TlsToolkitStandalone tls-toolkit standalone completed successfully
hw12203:...assembly/target/nifi-toolkit-1.1.0-SNAPSHOT-bin/nifi-toolkit-1.1.0-SNAPSHOT (pr1165) alopresto
🔓 1424s @ 19:52:14 $ ll localhost/
total 40
drwx------   5 alopresto  staff   170B Nov 14 19:52 ./
drwxr-xr-x  11 alopresto  staff   374B Nov 14 19:52 ../
-rw-------   1 alopresto  staff   3.4K Nov 14 19:52 keystore.pkcs12
-rw-------   1 alopresto  staff   8.5K Nov 14 19:52 nifi.properties
-rw-------   1 alopresto  staff   911B Nov 14 19:52 truststore.jks
hw12203:...assembly/target/nifi-toolkit-1.1.0-SNAPSHOT-bin/nifi-toolkit-1.1.0-SNAPSHOT (pr1165) alopresto
🔓 196s @ 19:55:31 $
```

File: nifi-commons/nifi-security-utils/src/main/java/org/apache/nifi/security/util/KeyStoreUtils.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    public static String getKeyStoreProvider(String keyStoreType) {
+        if (KeystoreType.PKCS12.toString().equalsIgnoreCase(keyStoreType)) {
+            return BouncyCastleProvider.PROVIDER_NAME;
+        }
+        return null;
+    }
+
+    public static KeyStore getKeyStore(String keyStoreType) throws KeyStoreException {
```
Comment (alopresto) [inline nifi-commons/nifi-security-utils/src/main/java/org/apache/nifi/security/util/KeyStoreUtils.java]: We should add Javadoc comments on these methods to indicate that they do not return a specific instance, rather a new empty instance similar to `KeyStore.getInstance()`.

**Comment by alopresto** [comment]
I think the Admin Guide should also get an update indicating that the TLS Toolkit will ignore user requests to use PKCS12 format for truststores.

**Comment by brosander** [comment]
Force push was a rebase for merge conflicts, addressing feedback now

File: nifi-commons/nifi-security-utils/src/main/java/org/apache/nifi/security/util/KeyStoreUtils.java (hunk 1 above)
Comment (brosander) [inline nifi-commons/nifi-security-utils/src/main/java/org/apache/nifi/security/util/KeyStoreUtils.java]: autocomplete fail, will address

File: nifi-docs/src/main/asciidoc/administration-guide.adoc [hunk 3]
```diff
@@ -174,6 +174,8 @@ TLS Generation Toolkit
 
 In order to facilitate the secure setup of NiFi, you can use the `tls-toolkit` command line utility to automatically generate the required keystores, truststore, and relevant configuration files. This is especially useful for securing multiple NiFi nodes, which can be a tedious and error-prone process.
 
+Note: JKS keyStores and trustStores are recommended for NiFi.  This tool allows the specification of other KeyStore types on the command line but will ignore a type of PKCS12 for use as the trustStore as that format has some compatibility issues between BouncyCastle and Oracle implementations.
```
Comment (alopresto) [inline nifi-docs/src/main/asciidoc/administration-guide.adoc]: To be consistent with the rest of the documentation, "keystore" and "truststore" should be capitalized as such.

Comment (brosander) [inline nifi-docs/src/main/asciidoc/administration-guide.adoc]: +1

**Comment by alopresto** [comment]
Performing final review (contrib-check, etc.).

**Comment by alopresto** [comment]
Review on hold for tonight -- logging issues. I will investigate if libraries or paths changed in the morning. 

```
hw12203:...assembly/target/nifi-toolkit-1.1.0-SNAPSHOT-bin/nifi-toolkit-1.1.0-SNAPSHOT (pr1165) alopresto
🔓 35s @ 18:59:45 $ ./bin/tls-toolkit.sh standalone -n 'localhost' -T PKCS12 -P password -S password
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/Users/alopresto/Workspace/nifi/nifi-toolkit/nifi-toolkit-assembly/target/nifi-toolkit-1.1.0-SNAPSHOT-bin/nifi-toolkit-1.1.0-SNAPSHOT/lib/logback-classic-1.1.3.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/Users/alopresto/Workspace/nifi/nifi-toolkit/nifi-toolkit-assembly/target/nifi-toolkit-1.1.0-SNAPSHOT-bin/nifi-toolkit-1.1.0-SNAPSHOT/lib/slf4j-log4j12-1.7.12.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [ch.qos.logback.classic.util.ContextSelectorStaticBinder]
hw12203:...assembly/target/nifi-toolkit-1.1.0-SNAPSHOT-bin/nifi-toolkit-1.1.0-SNAPSHOT (pr1165) alopresto
🔓 20s @ 19:00:06 $
```

**Comment by alopresto** [comment]
The logging issue was resolved by [NIFI-3049](https://issues.apache.org/jira/browse/NIFI-3049) and [PR 1237](https://github.com/apache/nifi/pull/1237). 

Verified `contrib-check` and all tests pass. Ran toolkit and logging output for PKCS12 truststore type is correct. Then ran application and was able to connect using client certificate as per usual. 

```
hw12203:...assembly/target/nifi-toolkit-1.1.0-SNAPSHOT-bin/nifi-toolkit-1.1.0-SNAPSHOT (pr1165) alopresto
🔓 46s @ 16:23:25 $ ./bin/tls-toolkit.sh standalone -n 'localhost' -T PKCS12 -P password -S password
2016/11/16 16:29:40 INFO [main] org.apache.nifi.toolkit.tls.commandLine.BaseCommandLine: Command line argument --keyStoreType=PKCS12 only applies to keystore, recommended truststore type of JKS unaffected.
2016/11/16 16:29:40 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandaloneCommandLine: No nifiPropertiesFile specified, using embedded one.
2016/11/16 16:29:41 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandalone: Running standalone certificate generation with output directory ../nifi-toolkit-1.1.0-SNAPSHOT
2016/11/16 16:29:41 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandalone: Generated new CA certificate ../nifi-toolkit-1.1.0-SNAPSHOT/nifi-cert.pem and key ../nifi-toolkit-1.1.0-SNAPSHOT/nifi-key.key
2016/11/16 16:29:41 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandalone: Writing new ssl configuration to ../nifi-toolkit-1.1.0-SNAPSHOT/localhost
2016/11/16 16:29:42 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandalone: Successfully generated TLS configuration for localhost 1 in ../nifi-toolkit-1.1.0-SNAPSHOT/localhost
2016/11/16 16:29:42 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandalone: No clientCertDn specified, not generating any client certificates.
2016/11/16 16:29:42 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandalone: tls-toolkit standalone completed successfully
hw12203:...assembly/target/nifi-toolkit-1.1.0-SNAPSHOT-bin/nifi-toolkit-1.1.0-SNAPSHOT (pr1165) alopresto
🔓 377s @ 16:29:43 $ ll localhost/
total 40
drwx------   5 alopresto  staff   170B Nov 16 16:29 ./
drwxr-xr-x  11 alopresto  staff   374B Nov 16 16:29 ../
-rw-------   1 alopresto  staff   3.4K Nov 16 16:29 keystore.pkcs12
-rw-------   1 alopresto  staff   8.6K Nov 16 16:29 nifi.properties
-rw-------   1 alopresto  staff   911B Nov 16 16:29 truststore.jks
hw12203:...assembly/target/nifi-toolkit-1.1.0-SNAPSHOT-bin/nifi-toolkit-1.1.0-SNAPSHOT (pr1165) alopresto
🔓 17s @ 16:30:01 $
```

Squashed, merged, and closed.

**Pull request closed without merging**